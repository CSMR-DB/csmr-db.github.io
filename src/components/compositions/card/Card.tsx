import React, { ReactNode } from 'react'
import styled, {
  StyledComponent,
  DefaultTheme,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { Tween } from 'react-gsap'

export interface ICardProps {
  theme?: DefaultTheme
  children: ReactNode
  index?: number
}

export const StyledCard: StyledComponent<
  'article',
  any,
  ICardProps,
  never
> = styled.article`
  position: relative;
  height: 100%;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  box-shadow: 0 0.075rem 0.075rem rgba(0, 0, 0, 0.1),
    0 0.125rem 0.125rem rgba(0, 0, 0, 0.25);
  transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);

  & header img {
    transition: all 0.2s ease-in-out !important;
  }

  & h1 {
    font-size: 1.1rem;
    margin: 0;
  }

  & h2 {
    font-size: 1rem;
    font-style: italic;
    margin: 0;
    color: #aaa;
    margin-bottom: 1rem;
  }

  & p {
    text-align: left;
  }

  ${({ theme }: ICardProps): FlattenSimpleInterpolation => css`
    background: ${theme!.palette.light.normal};

    @media ${theme!.breakpoints.min.smartphone} {
      text-align: justify;
    }

    @media ${theme!.breakpoints.min.desktop} {
      &:hover {
        box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.25),
          0 0.25rem 0.25rem rgba(0, 0, 0, 0.2);

        & header img {
          transform: scale(1.1);
        }
      }
    }
  `}
`

export function Card({ children, index = 0 }: ICardProps): JSX.Element {
  return (
    <Tween
      from={{ scale: 0.95, opacity: 0 }}
      duration={0.5}
      delay={0.25 + index * 0.125}
      ease="back"
    >
      <StyledCard>{children}</StyledCard>
    </Tween>
  )
}
