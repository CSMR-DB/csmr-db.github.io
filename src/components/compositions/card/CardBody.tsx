import styled, { StyledComponent } from 'styled-components'

export const CardBody: StyledComponent<'main', any, {}, never> = styled.main`
  position: relative;
  padding: 2em;
  grid-area: 'body';
`
