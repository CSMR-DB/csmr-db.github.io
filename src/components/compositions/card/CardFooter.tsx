import styled, {
  StyledComponent,
  FlattenSimpleInterpolation,
  css,
} from 'styled-components'

interface ICardFooterProps {
  $style?: boolean
}

export const CardFooter: StyledComponent<
  'footer',
  any,
  ICardFooterProps,
  never
> = styled.footer`
  ${({ $style: style = false }: ICardFooterProps): FlattenSimpleInterpolation =>
    style
      ? css`
          background: #eee;
          padding: 2rem;
          border-top: 1px solid #ddd;
          grid-area: footer;
        `
      : css`
          padding: 0 2rem 2rem;
        `}
`
