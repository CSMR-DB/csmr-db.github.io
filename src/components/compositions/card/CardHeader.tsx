import styled, { StyledComponent } from 'styled-components'

export const CardHeader: StyledComponent<
  'header',
  any,
  {},
  never
> = styled.header`
  grid-area: 'header';
`
