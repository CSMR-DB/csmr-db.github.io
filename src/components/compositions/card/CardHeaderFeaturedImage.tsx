import styled, {
  StyledComponent,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { MaxHeightProperty } from 'csstype'

interface ICardHeaderFeaturedImageProps {
  maxHeight?: MaxHeightProperty<1>
}

export const CardHeaderFeaturedImage: StyledComponent<
  'div',
  any,
  ICardHeaderFeaturedImageProps,
  never
> = styled.div`
  overflow: hidden;

  ${({
    maxHeight = '18rem',
  }: ICardHeaderFeaturedImageProps): FlattenSimpleInterpolation => css`
    max-height: ${maxHeight};
    min-height: ${maxHeight};

    & > * {
      max-height: ${maxHeight};
      min-height: ${maxHeight};
    }
  `}
`
