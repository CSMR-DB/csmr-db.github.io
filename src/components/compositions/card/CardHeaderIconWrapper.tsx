import styled, {
  StyledComponent,
  css,
  FlattenSimpleInterpolation,
} from 'styled-components'
import { BackgroundColorProperty } from 'csstype'

interface ICardHeaderIconWrapperProps {
  $backgroundColor: BackgroundColorProperty
}

export const CardHeaderIconWrapper: StyledComponent<
  'div',
  any,
  ICardHeaderIconWrapperProps,
  never
> = styled.div`
  border-radius: 999px;
  overflow: hidden;
  width: 4em;
  padding: 0.5em;

  ${({
    $backgroundColor: backgroundColor = '#DDD',
  }: ICardHeaderIconWrapperProps): FlattenSimpleInterpolation => css`
    background: ${backgroundColor};
  `}
`
