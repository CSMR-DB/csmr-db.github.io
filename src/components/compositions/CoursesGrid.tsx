import React from 'react'

import { ICourse } from '../../data/courses'

import { Flex } from '../Flex'
import { Grid } from '../Grid'
import { Card } from './card/Card'
import { CardHeader } from './card/CardHeader'
import { Underline } from '../Underline'
import { ColoredText } from '../ColoredText'
import { CardBody } from './card/CardBody'
import { Paddie } from '../Paddie'

export interface ICoursesGridProps {
  courses: [ICourse['category'], ICourse[]][]
}

export function CoursesGrid({ courses }: ICoursesGridProps): JSX.Element {
  return (
    <Flex $direction={'column'} $justifyContent={'center'}>
      <Grid $columns={6} $alignItems={'baseline'}>
        {courses.map(
          (
            courseGroup: [ICourse['category'], ICourse[]],
            i: number
          ): JSX.Element => (
            <Card key={i}>
              <CardHeader>
                <Paddie>
                  <Underline>
                    <h2>{courseGroup[0]}</h2>
                  </Underline>
                </Paddie>
              </CardHeader>
              <CardBody>
                {courseGroup[1].map(
                  (course: ICourse, j: number): JSX.Element => (
                    <h5 key={j}>
                      <ColoredText color={course.color ? 'black' : '#666'}>
                        {'‚ù≠'} {course.name}
                      </ColoredText>
                    </h5>
                  )
                )}
              </CardBody>
            </Card>
          )
        )}
      </Grid>
      <div>
        <p>
          * Not intrinsically about programming, I just made it so. ** Not a
          full on course with credits, but a support course to provide extra
          knowledge about Adobe software.
        </p>
      </div>
    </Flex>
  )
}
