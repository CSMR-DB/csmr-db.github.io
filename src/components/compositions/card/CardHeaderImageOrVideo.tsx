import React from 'react'
import Image from 'gatsby-image'
import { MaxHeightProperty } from 'csstype'

import { FileChildImageSharp } from '../../../types/graphql.types'
import {
  AnyExoticRefComponent,
  AnyExoticRefTargets,
} from '../../../types/react.types'
import { StaticDataManager } from '../../../data/StaticDataManager'

import { CardHeaderFeaturedImage } from './CardHeaderFeaturedImage'
import { DynamicImage } from '../DynamicImage'
import { StyledYoutube } from '../../StyledYoutube'

interface IImageOrVideo {
  image?: string | FileChildImageSharp
  video?: string
  maxHeight?: MaxHeightProperty<1>
}

export const CardHeaderImageOrVideo: AnyExoticRefComponent<IImageOrVideo> = React.forwardRef(
  (
    { image, video, maxHeight }: IImageOrVideo,
    ref: AnyExoticRefTargets
  ): JSX.Element => (
    <CardHeaderFeaturedImage $maxHeight={maxHeight} ref={ref}>
      {image && typeof image === 'string' ? (
        <DynamicImage
          path={image}
          dynamicImages={StaticDataManager.dynamicImages}
        />
      ) : (
        (image && typeof image !== 'string' && (
          <Image fluid={image.childImageSharp.fluid} />
        )) ||
        (video && (
          <StyledYoutube
            opts={{
              width: '960',
              height: '290',
            }}
            videoId={video.replace('https://www.youtube.com/embed/', '')}
          ></StyledYoutube>
        ))
      )}
    </CardHeaderFeaturedImage>
  )
)
