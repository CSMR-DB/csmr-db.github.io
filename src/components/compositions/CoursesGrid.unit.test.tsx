import React, { Component } from 'react'
import { ThemeProvider } from 'styled-components'
import { create, ReactTestRenderer } from 'react-test-renderer'
import { mount, ReactWrapper } from 'enzyme'

import { StaticDataManager } from '../../data/StaticDataManager'
import { groupBy } from '../../utils/groupBy'

import { CoursesGrid } from './CoursesGrid'
import { Card } from './card/Card'

jest.mock('react-gsap')

describe('CoursesGrid', (): void => {
  it('should render a table-like set of grouped courses', (): void => {
    const comp: ReactTestRenderer = create(
      <ThemeProvider theme={StaticDataManager.theme}>
        <CoursesGrid
          courses={[...groupBy(StaticDataManager.courses, 'category')]}
        />
      </ThemeProvider>
    )

    expect(comp).toMatchSnapshot()
  })

  it('should render 1 Card element', (): void => {
    const comp: ReactWrapper<any, Readonly<{}>, Component<{}, {}, any>> = mount(
      <ThemeProvider theme={StaticDataManager.theme}>
        <CoursesGrid
          courses={[...groupBy(StaticDataManager.courses, 'category')]}
        />
      </ThemeProvider>
    )

    expect(comp.find(Card).length).toEqual(1)
  })
})
