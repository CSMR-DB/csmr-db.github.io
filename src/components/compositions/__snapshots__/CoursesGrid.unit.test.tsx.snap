// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursesGrid should render a table-like set of grouped courses 1`] = `
.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  height: 100%;
  width: 100%;
  -webkit-transition: all 0.25s ease-in-out;
  transition: all 0.25s ease-in-out;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-wrap: initial;
  -ms-flex-wrap: initial;
  flex-wrap: initial;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  display: grid;
  margin: 0 auto;
  height: auto;
  max-width: auto;
  gap: calc(1rem / 2);
  grid-template-columns: repeat(auto-fill,minmax(16rem,1fr));
  grid-template-rows: auto;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c2 {
  position: relative;
  height: 100%;
  border-radius: 0.25rem;
  display: grid;
  grid-template-areas: 'header' 'body' 'footer';
  grid-template-rows: auto 1fr auto;
  overflow: hidden;
  box-shadow: 0 0.075rem 0.075rem rgba(0,0,0,0.1), 0 0.125rem 0.125rem rgba(0,0,0,0.25);
  -webkit-transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  transition: all 0.3s cubic-bezier(0.25,0.8,0.25,1);
  background: #fff;
}

.c2 header img {
  -webkit-transition: all 0.2s ease-in-out !important;
  transition: all 0.2s ease-in-out !important;
}

.c2 h1 {
  font-size: 1.1rem;
  margin: 0;
}

.c2 h2 {
  font-size: 1rem;
  font-style: italic;
  margin: 0;
  color: #aaa;
  margin-bottom: 1rem;
}

.c2 p {
  text-align: left;
}

.c3 {
  grid-area: 'header';
}

.c5 {
  width: 100%;
  border-bottom: 1px dashed #ddd;
}

.c7 {
  color: #000;
}

.c6 {
  grid-area: 'body';
  position: relative;
  padding: 2rem;
}

.c4 {
  padding: 2rem;
}

@media only screen and (min-width:600px) {
  .c1 {
    grid-template-columns: repeat(6,1fr);
    gap: 1rem;
  }
}

@media only screen and (min-width:600px) {
  .c2 {
    text-align: justify;
  }
}

@media only screen and (min-width:1200px) {
  .c2:hover {
    box-shadow: 0 0.125rem 0.25rem rgba(0,0,0,0.25), 0 0.25rem 0.25rem rgba(0,0,0,0.2);
  }

  .c2:hover header img {
    -webkit-transform: scale(1.1);
    -ms-transform: scale(1.1);
    transform: scale(1.1);
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <article
      className="c2"
    >
      <header
        className="c3"
      >
        <div
          className="c4"
        >
          <div
            className="c5"
          >
            <h2>
              Development
            </h2>
          </div>
        </div>
      </header>
      <main
        className="c6"
      >
        <h5>
          <span
            className="c7"
          >
            ‚ù≠
             
            WebProgramming
          </span>
        </h5>
      </main>
    </article>
  </div>
  <div>
    <p>
      * Not intrinsically about programming, I just made it so. ** Not a full on course with credits, but a support course to provide extra knowledge about Adobe software.
    </p>
  </div>
</div>
`;
